rpcs3/Gui/gamelistframe.cpp

```diff
@@ -1,18 +1,576 @@
+#include <QTimer>
+
+static const std::string m_class_name = "GameViewer";
+
+// Auxiliary classes
+class sortGameData
+{
+	int sortColumn;
+	bool sortAscending;
+
+public:
+	sortGameData(u32 column, bool ascending) : sortColumn(column), sortAscending(ascending) {}
+	bool operator()(const GameInfo& game1, const GameInfo& game2) const
+	{
+		// Note that the column index has to match the appropriate GameInfo member
+		switch (sortColumn - 1) // skip *icon* column
+		{
+		case 0: return sortAscending ? (game1.name < game2.name) : (game1.name > game2.name);
+		case 1: return sortAscending ? (game1.serial < game2.serial) : (game1.serial > game2.serial);
+		case 2: return sortAscending ? (game1.fw < game2.fw) : (game1.fw > game2.fw);
+		case 3: return sortAscending ? (game1.app_ver < game2.app_ver) : (game1.app_ver > game2.app_ver);
+		case 4: return sortAscending ? (game1.category < game2.category) : (game1.category > game2.category);
+		case 5: return sortAscending ? (game1.root < game2.root) : (game1.root > game2.root);
+		default: return false;
+		}
+	}
+};
 
 GameListFrame::GameListFrame(QWidget *parent) : QDockWidget(tr("Game List"), parent)
 {
 	gameList = new QTableWidget(this);
+	gameList->setSelectionBehavior(QAbstractItemView::SelectRows);
+	gameList->verticalHeader()->setVisible(false);
+
 	gameList->setColumnCount(7);
 	gameList->setHorizontalHeaderItem(0, new QTableWidgetItem(tr("Icon")));
 	gameList->setHorizontalHeaderItem(1, new QTableWidgetItem(tr("Name")));
 	gameList->setHorizontalHeaderItem(2, new QTableWidgetItem(tr("Serial")));
 	gameList->setHorizontalHeaderItem(3, new QTableWidgetItem(tr("FW")));
-	gameList->setHorizontalHeaderItem(4, new QTableWidgetItem(tr("Version")));
+	gameList->setHorizontalHeaderItem(4, new QTableWidgetItem(tr("App version")));
 	gameList->setHorizontalHeaderItem(5, new QTableWidgetItem(tr("Category")));
 	gameList->setHorizontalHeaderItem(6, new QTableWidgetItem(tr("Path")));
 
 	setWidget(gameList);
+
+	gameList->setContextMenuPolicy(Qt::CustomContextMenu);
+	connect(gameList, &QTableWidget::customContextMenuRequested, this, &GameListFrame::ShowContextMenu);
+	connect(gameList, &QTableWidget::doubleClicked, this, &GameListFrame::doubleClickedSlot);
+	connect(gameList->horizontalHeader(), &QHeaderView::sectionClicked, this, &GameListFrame::OnColClicked);
+
+	Refresh();
+
+	m_sortColumn = 1; // sort by name by default
+
+	// Refresh game list every 10 seconds.
+	QTimer *timer = new QTimer(this);
+	connect(timer, &QTimer::timeout, this, &GameListFrame::Refresh);
+	timer->start(10000);
+}
+
+GameListFrame::~GameListFrame()
+{
+
+}
+
+void GameListFrame::OnColClicked(int col)
+{
+	if (col == m_sortColumn)
+		m_sortAscending ^= true;
+	else
+		m_sortAscending = true;
+	m_sortColumn = col;
+
+	// Sort entries, update columns and refresh the panel
+	std::sort(m_game_data.begin(), m_game_data.end(), sortGameData(m_sortColumn, m_sortAscending));
+	m_columns.Update(m_game_data);
+	ShowData();
```

Comment:
Replace these three lines with Refresh(); in order to get rid of doublesort issue